odoo/addons/base/models/ir_attachment.py

```diff
@@ -318,6 +320,30 @@ def _auto_init(self):
                            self._table, ['res_model', 'res_id'])
         return res
 
+    def _register_hook(self):
+        super()._register_hook()
+
+        m2ms_to_self = (
+            field
+            for model in self.env.values()
+            for field in model._fields.values()
+            if field.type == 'many2many'
+            if field.comodel_name == 'ir.attachment'
+            if field.store
+        )
+
+        for f in m2ms_to_self:
+            self.env.cr.execute(sql.SQL("""
+            UPDATE ir_attachment a
+            SET res_model = %s, res_id = r.{id1}
+            FROM {rel} r
+            WHERE a.id = r.{id2} AND a.res_id = 0
```

Comment:
Woah, this may change attachments in a "random" way.  If the many2many is really such a relation, a single attachment can be related to two records or more.  This update will set the `res_id` of the attachment to any of them, in a nondeterministic way.